File: QuasselDroidNG/src/main/java/de/kuschku/libquassel/primitives/serializers/MessageSerializer.java
package de.kuschku.libquassel.primitives.serializers;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.joda.time.DateTime;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.ByteChannel;

import de.kuschku.libquassel.message.Message;

import static de.kuschku.util.AndroidAssert.*;

public class MessageSerializer implements PrimitiveSerializer<Message> {
    @NonNull
    private static final MessageSerializer serializer = new MessageSerializer();

    private MessageSerializer() {
    }

    @NonNull
    public static MessageSerializer get() {
        return serializer;
    }

    @Override
    public void serialize(@NonNull ByteChannel channel, @NonNull Message data) throws IOException {
        IntSerializer.get().serialize(channel, data.messageId);
        IntSerializer.get().serialize(channel, (int) (data.time.getMillis() / 1000));
        IntSerializer.get().serialize(channel, data.type.value);
        ByteSerializer.get().serialize(channel, data.flags.flags);
        BufferInfoSerializer.get().serialize(channel, data.bufferInfo);
        ByteArraySerializer.get().serialize(channel, data.sender);
        ByteArraySerializer.get().serialize(channel, data.content);
    }

    @Nullable
    @Override
    public Message deserialize(@NonNull final ByteBuffer buffer) throws IOException {
        String sender = ByteArraySerializer.get().deserialize(buffer);
        String message = ByteArraySerializer.get().deserialize(buffer);

        assertNotNull(sender);
        assertNotNull(message);

        return new Message(
                IntSerializer.get().deserialize(buffer),
                new DateTime(((long) IntSerializer.get().deserialize(buffer)) * 1000),
                Message.Type.fromId(IntSerializer.get().deserialize(buffer)),
                new Message.Flags(ByteSerializer.get().deserialize(buffer)),
                BufferInfoSerializer.get().deserialize(buffer),
                sender,
                message
        );
    }
}

File: protocol/src/main/java/de/kuschku/libquassel/protocol/serializers/primitive/MessageSerializer.kt
/*
 * Quasseldroid - Quassel client for Android
 *
 * Copyright (c) 2020 Janne Mareike Koschinski
 * Copyright (c) 2020 The Quassel Project
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 3 as published
 * by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.libquassel.protocol.serializers.primitive

import de.kuschku.bitflags.of
import de.kuschku.bitflags.toBits
import de.kuschku.libquassel.protocol.features.FeatureSet
import de.kuschku.libquassel.protocol.features.QuasselFeature
import de.kuschku.libquassel.protocol.io.ChainedByteBuffer
import de.kuschku.libquassel.protocol.types.Message
import de.kuschku.libquassel.protocol.types.MessageFlag
import de.kuschku.libquassel.protocol.types.MessageType
import de.kuschku.libquassel.protocol.variant.QuasselType
import org.threeten.bp.Instant
import java.nio.ByteBuffer

object MessageSerializer : QuasselSerializer<Message> {
  override val quasselType: QuasselType = QuasselType.Message
  override val javaType: Class<out Message> = Message::class.java

  override fun serialize(buffer: ChainedByteBuffer, data: Message, featureSet: FeatureSet) {
    MsgIdSerializer.serialize(buffer, data.messageId, featureSet)
    if (featureSet.hasFeature(QuasselFeature.LongTime)) {
      LongSerializer.serialize(buffer, data.time.toEpochMilli(), featureSet)
    } else {
      IntSerializer.serialize(buffer, data.time.epochSecond.toInt(), featureSet)
    }
    UIntSerializer.serialize(buffer, data.type.toBits(), featureSet)
    UByteSerializer.serialize(buffer, data.flag.toBits().toUByte(), featureSet)
    BufferInfoSerializer.serialize(buffer, data.bufferInfo, featureSet)
    StringSerializerUtf8.serialize(buffer, data.sender, featureSet)
    if (featureSet.hasFeature(QuasselFeature.SenderPrefixes)) {
      StringSerializerUtf8.serialize(buffer, data.senderPrefixes, featureSet)
    }
    if (featureSet.hasFeature(QuasselFeature.RichMessages)) {
      StringSerializerUtf8.serialize(buffer, data.realName, featureSet)
      StringSerializerUtf8.serialize(buffer, data.avatarUrl, featureSet)
    }
    StringSerializerUtf8.serialize(buffer, data.content, featureSet)
  }

  override fun deserialize(buffer: ByteBuffer, featureSet: FeatureSet): Message {
    return Message(
      messageId = MsgIdSerializer.deserialize(buffer, featureSet),
      time = if (featureSet.hasFeature(QuasselFeature.LongTime))
        Instant.ofEpochMilli(LongSerializer.deserialize(buffer, featureSet))
      else
        Instant.ofEpochSecond(IntSerializer.deserialize(buffer, featureSet).toLong()),
      type = MessageType.of(UIntSerializer.deserialize(buffer, featureSet)),
      flag = MessageFlag.of(
        UByteSerializer.deserialize(buffer, featureSet).toUInt()
      ),
      bufferInfo = BufferInfoSerializer.deserialize(buffer, featureSet),
      sender = StringSerializerUtf8.deserialize(buffer, featureSet) ?: "",
      senderPrefixes = if (featureSet.hasFeature(QuasselFeature.SenderPrefixes))
        StringSerializerUtf8.deserialize(buffer, featureSet) ?: "" else "",
      realName = if (featureSet.hasFeature(QuasselFeature.RichMessages))
        StringSerializerUtf8.deserialize(buffer, featureSet) ?: "" else "",
      avatarUrl = if (featureSet.hasFeature(QuasselFeature.RichMessages))
        StringSerializerUtf8.deserialize(buffer, featureSet) ?: "" else "",
      content = StringSerializerUtf8.deserialize(buffer, featureSet) ?: ""
    )
  }
}
